core/utils/deferable_writer_closer.go

```diff
@@ -0,0 +1,62 @@
+package utils
+
+import (
+	"io"
+	"sync"
+)
+
+// DeferableWriterCloser is to be used in leiu of defer'ing
+// Close on an io.WriterClose (// For more background see https://www.joeshaw.org/dont-defer-close-on-writable-files/)
+// Callers should *both*
+// explicitly call Close and check for errors when done with the underlying writerclose
+// *and* defer the Close() to handle returns before the explicit close
+//
+// For example rather than
+// f, err := os.Create(...)
+// if err != nil {...}
+// defer f.Close()
+// f.Write(...)
+// ...
+// return
+//
+// do
+//
+// f, err := os.Create(...)
+// if err != nil {...}
+// wc := NewDeferableWriterCloser(f)
+// defer wc.Close()
+// wc.Write(...)
+// ...
+// err = wc.Close()
+// if err != nil {...}
+// return
```

Comment:
nit/ could indent these code blocks
new go docs has more link support too, if you wish: https://go.dev/doc/comment#links

Reply:
done